### PR TITLE
RWA-237 Refactor task endpoint

### PR DESCRIPTION
**Pull request by lewis150193**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-237

### Change description ###
Refactored workflow-api controller for send message and evaluate dmn

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x ] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    public void setUp() {
+        caseId = UUID.randomUUID().toString();
+        serviceAuthorizationToken =
+            authorizationHeadersProvider
+                .getAuthorizationHeaders()
+                .getValue(SERVICE_AUTHORIZATION);
+    }
+
+    @Test
+    public void should_not_allow_requests_without_valid_service_authorisation_and_return_403_response_code() {
```
Comment (raghera) [inline src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java:45]: I know some of this is from what Chris initially did.  Are there some more negative tests covered elsewhere?  E.g. if you provide an invalid `id`.

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    public void setUp() {
+        caseId = UUID.randomUUID().toString();
+        serviceAuthorizationToken =
+            authorizationHeadersProvider
+                .getAuthorizationHeaders()
+                .getValue(SERVICE_AUTHORIZATION);
+    }
+
+    @Test
+    public void should_not_allow_requests_without_valid_service_authorisation_and_return_403_response_code() {
```
Comment (raghera) [inline src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java:44]: similar comment to above around negative tests.

**Review by iguannago — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientService.java [hunk 3]
```diff
@@ -24,48 +19,23 @@ public TaskClientService(@Autowired CamundaClient camundaClient,
...
-            serviceDetails.getJurisdiction(),
-            serviceDetails.getCaseType(),
-            caseId,
-            taskToCreate.getTask(),
-            taskToCreate.getGroup(),
-            dueDate,
-            taskToCreate.getName()
-        );
-        camundaClient.sendMessage(
+    public List<Map<String,DmnValue>> evaluate(EvaluateDmnRequest evaluateDmnRequest, String id) {
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientService.java:29]: It seems the id is not used here, right?

**Comment by iguannago** [comment]
Where is the calculate date logic used in master when creating the task?

File: src/main/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientService.java (hunk 3 above)
Comment (lewis150193) [reply src/main/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientService.java:29]: Yh the id is only used for the evaluateDmn endpoint the sendMessage should work via the messageName

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java (hunk 1 above)
Comment (lewis150193) [reply src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java:45]: I have added tests in for both now which handle these situations

File: src/main/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientService.java (hunk 3 above)
Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientService.java:29]: ok, so then can we remove it? If it's not used...

Comment (iguannago) [reply src/main/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientService.java:29]: Or maybe you mean it's planned to be used later on.

**Comment by iguannago** [comment]
@lewis150193 , I pushed a very little change to authorise the case-event-handler service. Otherwise, it's not going to work either on Preview or AAT.

**Review by iguannago — APPROVED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            .pathParam("key","getTask_IA_Asylum")
+            .basePath("/workflow/decision-definition/{key}/evaluate")
+            .when()
+            .post()
+            .then()
+            .statusCode(HttpStatus.FORBIDDEN_403);
+    }
+
+    @Test
+    public void should_evaluate_and_return_dmn_results() {
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java:61]: As this is an edge case, could we move this into Integration test & mock the expectation from the camunda standalone service?

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            .baseUri(testUrl)
+            .basePath("/workflow/message")
+            .when()
+            .post()
+            .then()
+            .statusCode(HttpStatus.NO_CONTENT_204);
+    }
+
+    @Test
+    public void should_not_be_able_to_post_as_message_does_not_exist() {
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java:75]: Candidate for Integration Test?

File: src/integrationTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/CreateTaskTest.java [hunk 6]
```diff
@@ -58,118 +42,59 @@ void setUp() {
         when(authTokenGenerator.generate()).thenReturn(BEARER_SERVICE_TOKEN);
     }
 
-    @DisplayName("Should create task with default due date and 201 response")
+    @DisplayName("Should evaluate a DMN and return a 200")
     @Test
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/CreateTaskTest.java:46]: the dueDate tests covered the AC's for the original ticket? Where are the tests moved too?

File: src/test/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientServiceTest.java [hunk 7]
```diff
@@ -2,156 +2,64 @@
...
-        transition = new Transition("startState", "eventName", "endState");
-        dmnRequest = new DmnRequest<>(new GetTaskDmnRequest(
-            dmnStringValue(transition.getEventId()),
-            dmnStringValue(transition.getPostState())
-        ));
-        serviceDetails = new ServiceDetails("jurisdiction", "caseType");
-
+        evaluateDmnRequest = new EvaluateDmnRequest(Map.of("name",DmnValue.dmnStringValue("test")));
         when(authTokenGenerator.generate()).thenReturn(BEARER_SERVICE_TOKEN);
     }
```
Comment (BenNzewi) [inline src/test/java/uk/gov/hmcts/reform/waworkflowapi/external/taskservice/TaskClientServiceTest.java:38]: Working days test? Where have these been moved too?

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            .relaxedHTTPSValidation()
+            .header(SERVICE_AUTHORIZATION, serviceAuthorizationToken)
+            .contentType(APPLICATION_JSON_VALUE)
+            .body(new EvaluateDmnRequest(mockVariables())).log().body()
+            .baseUri(testUrl)
+            .pathParam("key",WA_TASK_INITIATION_IA_ASYLUM)
+            .basePath("/workflow/decision-definition/key/{key}/evaluate")
+            .when()
+            .post()
+            .prettyPeek()
```
Comment (alexbaeza) [inline src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java:62]: I think we should not output to the console

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java [hunk 9]
```diff
@@ -109,26 +106,36 @@ public void transition_creates_atask_with_default_due_date() {
...
-            .statusCode(HttpStatus.CREATED_201);
+            .statusCode(HttpStatus.NO_CONTENT_204);
 
         Object taskId = given()
             .header(SERVICE_AUTHORIZATION, serviceAuthorizationToken)
             .contentType(APPLICATION_JSON_VALUE)
             .baseUri(camundaUrl)
             .basePath("/task")
-            .param("processVariables", "caseId_eq_" + caseId)
+            .param("processVariables", "caseReference_eq_" + caseReference)
```
Comment (alexbaeza) [inline src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java:138]: according to the task-configuration this should be caseId
The service is overdue a cleanup and refactoring but it's defined here:
https://github.com/hmcts/wa-task-configuration-api/blob/4fecb40f7e82a20281e6c13bf238d9e195b2f99c/src/main/java/uk/gov/hmcts/reform/wataskconfigurationapi/ccdmapping/ConfigureTaskService.java#L22

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java [hunk 10]
```diff
@@ -53,40 +43,47 @@ public void setUp() {
...
+            "processApplication",
+            "TCW"
+        );
 
         given()
             .relaxedHTTPSValidation()
             .header(SERVICE_AUTHORIZATION, serviceAuthorizationToken)
             .contentType(APPLICATION_JSON_VALUE)
-            .body(appealSubmittedCreateTaskRequest(caseId)).log().body()
+            .body(new SendMessageRequest("createTaskMessage", processVariables)).log().body()
```
Comment (alexbaeza) [inline src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java:73]: Minor suggestion for easier mantainance in the future "createTaskMessage" should be an enum as task-managament, see: https://github.com/hmcts/wa-task-management-api/blob/master/src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/domain/entities/camunda/CamundaMessage.java

File: src/testUtils/java/uk/gov/hmcts/reform/waworkflowapi/SpringBootFunctionalBaseTest.java [hunk 11]
```diff
@@ -2,12 +2,21 @@
...
 import net.serenitybdd.junit.spring.integration.SpringIntegrationSerenityRunner;
 import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.test.context.ActiveProfiles;
 
 @RunWith(SpringIntegrationSerenityRunner.class)
 @SpringBootTest
-@ActiveProfiles("functional")
+@ActiveProfiles(profiles = {"local","functional"})
```
Comment (alexbaeza) [inline src/testUtils/java/uk/gov/hmcts/reform/waworkflowapi/SpringBootFunctionalBaseTest.java:11]: do we need local to run on the pipeline too :) ?

Comment (iguannago) [reply src/testUtils/java/uk/gov/hmcts/reform/waworkflowapi/SpringBootFunctionalBaseTest.java:11]: removed.

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java (hunk 9 above)
Comment (iguannago) [reply src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java:138]: ok, I changed it because the contract message from CCD says caseReference. But yeah, I change it now.

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java (hunk 8 above)
Comment (iguannago) [reply src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/EvaluateDmnTest.java:62]: ok

File: src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java (hunk 10 above)
Comment (iguannago) [reply src/functionalTest/java/uk/gov/hmcts/reform/waworkflowapi/controllers/SendMessageTest.java:73]: yes, all these tests and logic will move in the next sprint. There is a ticket to move the dueDat to the case event handler. 
I have just put back the original tests and did not spend more than necessary to make it work since as we say it's going to move. Actually, the whole dueDate is subject to major changes, so probably is not worth the effort at this moment.

**Review by alexbaeza — APPROVED** [review]
